src/NMS.AMQP/Util/AmqpDestinationHelper.cs

```diff
@@ -0,0 +1,240 @@
+        {
+            if (address == null)
+                return null;
+
+            switch (typeByte)
+            {
+                case MessageSupport.JMS_DEST_TYPE_QUEUE:
+                    return new NmsQueue(address);
+                case MessageSupport.JMS_DEST_TYPE_TOPIC:
+                    return new NmsTopic(address);
+                case MessageSupport.JMS_DEST_TYPE_TEMP_QUEUE:
+                    NmsTemporaryQueue temporaryQueue = new NmsTemporaryQueue(new CustomIdGenerator(true, address).GenerateId());
+                    temporaryQueue.Address = address;
+                    return temporaryQueue;
+                case MessageSupport.JMS_DEST_TYPE_TEMP_TOPIC:
+                    NmsTemporaryTopic temporaryTopic = new NmsTemporaryTopic(new CustomIdGenerator(true, address).GenerateId());
+                    temporaryTopic.Address = address;
+                    return temporaryTopic;
+            }
+
+            // fall back to a Queue Destination since we need a real NMS destination
+            return new NmsQueue(address);
+        }
+
+        private static IDestination CreateDestination(string address, IDestination consumerDestination, bool useConsumerDestForTypeOnly)
+        {
+            if (address == null)
+            {
+                return useConsumerDestForTypeOnly ? null : consumerDestination;
+            }
+
+            if (consumerDestination.IsQueue)
+            {
+                if (consumerDestination.IsTemporary)
+                    return new NmsTemporaryQueue(new CustomIdGenerator(true, address).GenerateId()) { Address = address };
+                else
+                    return new NmsQueue(address);
+            }
+            else if (consumerDestination.IsTopic)
+            {
+                if (consumerDestination.IsTemporary)
+                    return new NmsTemporaryTopic(new CustomIdGenerator(true, address).GenerateId()) { Address = address };
+                else
+                    return new NmsTopic(address);
+            }
+
+            // fall back to a Queue Destination since we need a real NMS destination
+            return new NmsQueue(address);
+        }
+
+        public static void SetToAnnotationFromDestination(IDestination destination, MessageAnnotations annotations)
+            => SetAnnotationFromDestination(SymbolUtil.JMSX_OPT_DEST, destination, annotations);
+
+        public static void SetReplyToAnnotationFromDestination(IDestination destination, MessageAnnotations annotations)
+            => SetAnnotationFromDestination(SymbolUtil.JMSX_OPT_REPLY_TO, destination, annotations);
+
+        private static void SetAnnotationFromDestination(Symbol key, IDestination destination, MessageAnnotations annotations)
+        {
+            byte? typeValue = ToTypeAnnotation(destination);
+
+            if (typeValue == null)
+                annotations.Map.Remove(key);
+            else
+                annotations.Map.Add(key, typeValue);
```

Comment:
This throws an exception when a message object is reused to send multiple messages as adding a pre-existing key to the annotation map throws an exception.
```suggestion
                annotations.Map.[key] = typeValue;
```
Or
```suggestion
     else if (!annotations.Map.ContainsKey(key))
                annotations.Map.Add(key, typeValue);
     else
                annotations.Map.[key] = typeValue;
```

Reply:
Fixed.